File: 구름 알고리즘 미션/완전 탐색/문자열 나누기/문자열 나누기.js
// Run by Node.js
const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });

const input = [];

rl.on('line', function (line) {
	input.push(line);
}).on('close', function () {
	const N = +input[0];
	const S = input[1];
	
	let result = [];
	for(let i = 1 ; i < N-1 ; i++){
		for(let j = i+1 ; j < N ; j++){
			const p1 = S.slice(0, i);
			const p2 = S.slice(i, j);
			const p3 = S.slice(j);
			result.push([p1, p2, p3]);
		}
	}
	
	let flat = [];
	for(let r of result) flat.push(...r);
	const P = [...new Set(flat)].sort();
	
	let maxScore = 0;
	result.forEach(([p1, p2, p3]) => {
		let i = P.indexOf(p1), j = P.indexOf(p2), k = P.indexOf(p3);
		const score = i + j + k + 3;
		maxScore = score > maxScore ? score : maxScore;
	})
	
	console.log(maxScore);
	process.exit();
});

File: 구름 알고리즘 미션/완전 탐색/문자열 나누기/문자열 나누기.py
N = int(input())
S = input()

result = []
for i in range(1, N - 1):
    for j in range(i + 1, N):
        part1 = S[:i]
        part2 = S[i:j]
        part3 = S[j:]
        result.append([part1, part2, part3])

# 모든 부분문자열 수집 후 중복 제거하고 정렬
flat = []
for r in result:
    flat.extend(r)
P = sorted(set(flat))

# 점수 계산
max_score = 0
for part1, part2, part3 in result:
    i = P.index(part1)
    j = P.index(part2)
    k = P.index(part3)
    score = i + j + k + 3
    max_score = max(max_score, score)

print(max_score)
